apps/hash-ai-worker-ts/src/activities/flow-activities/research-entities-action/link-follower-agent.ts

```diff
@@ -0,0 +1,406 @@
+import type { SourceProvenance } from "@local/hash-graph-client";
+import { SourceType } from "@local/hash-graph-client";
+import dedent from "dedent";
+import { MetadataMode } from "llamaindex";
+
+import { getWebPageActivity } from "../../get-web-page-activity";
+import type { DereferencedEntityTypesByTypeId } from "../../infer-entities/inference-types";
+import { logger } from "../../shared/activity-logger";
+import type { DereferencedEntityType } from "../../shared/dereference-entity-type";
+import { getFlowContext } from "../../shared/get-flow-context";
+import { logProgress } from "../../shared/log-progress";
+import { stringify } from "../../shared/stringify";
+import { inferFactsFromText } from "../shared/infer-facts-from-text";
+import type { LocalEntitySummary } from "../shared/infer-facts-from-text/get-entity-summaries-from-text";
+import type { Fact } from "../shared/infer-facts-from-text/types";
+import { deduplicateEntities } from "./deduplicate-entities";
+import type { Link } from "./link-follower-agent/extract-links-from-content";
+import { extractLinksFromContent } from "./link-follower-agent/extract-links-from-content";
+import { filterAndRankTextChunksAgent } from "./link-follower-agent/filter-and-rank-text-chunks-agent";
+import { getLinkFollowerNextToolCalls } from "./link-follower-agent/get-link-follower-next-tool-calls";
+import { indexPdfFile } from "./link-follower-agent/llama-index/index-pdf-file";
+
+type ResourceToExplore = {
+  url: string;
+  descriptionOfExpectedContent: string;
+  exampleOfExpectedContent: string;
+  reason: string;
+};
+
+type LinkFollowerAgentInput = {
+  initialResource: ResourceToExplore;
+  task: string;
+  entityTypes: DereferencedEntityType[];
+  linkEntityTypes?: DereferencedEntityType[];
+};
+
+const isContentAtUrlPdfFile = async (params: { url: string }) => {
+  const { url } = params;
+  const urlHeadFetch = await fetch(url, { method: "HEAD" });
```

Comment:
We should wrap this in a `try / catch` otherwise some failures to fetch will tank the step with an unhandled exception

Reply:
Sure, done in https://github.com/hashintel/hash/pull/4649/commits/8f4cf5ae32788f8ce481b3abbb4c9130c5f83aa9